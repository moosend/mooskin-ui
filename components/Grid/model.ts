import { IBoxComponentProps } from '../Box/model';

export interface IGridProps extends IBoxComponentProps {
	children?: Array<React.ReactElement<IRowProps>> | React.ReactElement<IRowProps>;
}

export interface IRowProps extends IBoxComponentProps {
	children?: Array<React.ReactElement<IColProps>> | React.ReactElement<IColProps>;
}

export type ColWidth = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export interface IColWidths {
	lg?: ColWidth | 'hidden';
	md?: ColWidth | 'hidden';
	sm?: ColWidth | 'hidden';
	xs?: ColWidth | 'hidden';
}

export interface IColProps extends IColWidths, IBoxComponentProps {
	children?: any;
}
