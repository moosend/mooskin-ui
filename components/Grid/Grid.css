@import '../_utils/globals/screens.css';

.grid {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    flex-shrink: 1;
    box-sizing: border-box;
}

.grid::after, .row::after{
    clear: both;
}

.row{
    display: flex;
    flex-flow: row wrap;
    box-sizing: border-box;
}

.col{
    box-sizing: border-box;
}

.default {
    flex-grow: 1;
    flex-shrink: 1;
}

.hidden{
    display: none;
}

/* .activeLoader{
    display: flex;
    width: 100%;
    height: 100%;
    position: absolute;
    align-items: center;
    justify-content: center;
} */

.hiddenContent{
    visibility: hidden;
}

@media screen and (--large){

    .large-1 {flex: 1 1 8.33333333%}

    .large-2 {flex: 2 1 16.66666667%}

    .large-3 {flex: 3 1 25%}

    .large-4 {flex: 4 1 33.33333333%}

    .large-5 {flex: 5 1 41.66666667%}

    .large-6 {flex: 6 1 50%}

    .large-7 {flex: 7 1 58.33333333%}

    .large-8 {flex: 8 1 66.66666667%}

    .large-9 {flex: 9 1 75%}

    .large-10 {flex: 10 1 83.33333333%}

    .large-11 {flex: 11 1 91.66666667%}

    .large-12 {flex: 12 1 100%}

    .hidden-lg {display: none}
}

@media screen and (--medium){

    .medium-1 {flex: 1 1 8.33333333%}

    .medium-2 {flex: 2 1 16.66666667%}

    .medium-3 {flex: 3 1 25%}

    .medium-4 {flex: 4 1 33.33333333%}

    .medium-5 {flex: 5 1 41.66666667%}

    .medium-6 {flex: 6 1 50%}

    .medium-7 {flex: 7 1 58.33333333%}

    .medium-8 {flex: 8 1 66.66666667%}

    .medium-9 {flex: 9 1 75%}

    .medium-10 {flex: 10 1 83.33333333%}

    .medium-11 {flex: 11 1 91.66666667%}

    .medium-12 {flex: 12 1 100%}

    .hidden-md {display: none}
}


@media screen and (--small){

    .small-1 {flex: 1 1 8.33333333%}

    .small-2 {flex: 2 1 16.66666667%}

    .small-3 {flex: 3 1 25%}

    .small-4 {flex: 4 1 33.33333333%}

    .small-5 {flex: 5 1 41.66666667%}

    .small-6 {flex: 6 1 50%}

    .small-7 {flex: 7 1 58.33333333%}

    .small-8 {flex: 8 1 66.66666667%}

    .small-9 {flex: 9 1 75%}

    .small-10 {flex: 10 1 83.33333333%}

    .small-11 {flex: 11 1 91.66666667%}

    .small-12 {flex: 12 1 100%}

    .hidden-sm {display: none}
}

@media (--xSmall){

    .xSmall-1 {flex: 1 1 8.33333333%}

    .xSmall-2 {flex: 2 1 16.66666667%}

    .xSmall-3 {flex: 3 1 25%}

    .xSmall-4 {flex: 4 1 33.33333333%}

    .xSmall-5 {flex: 5 1 41.66666667%}

    .xSmall-6 {flex: 6 1 50%}

    .xSmall-7 {flex: 7 1 58.33333333%}

    .xSmall-8 {flex: 8 1 66.66666667%}

    .xSmall-9 {flex: 9 1 75%}

    .xSmall-10 {flex: 10 1 83.33333333%}

    .xSmall-11 {flex: 11 1 91.66666667%}

    .xSmall-12 {flex: 12 1 100%}  

    .hidden-xs {display: none}
}
