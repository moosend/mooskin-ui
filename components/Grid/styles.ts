import styled from 'styled-components';

// Models
import { ColWidth, IColProps, IColWidths, IGridProps, IRowProps } from './model';

// Components
import { Box } from '../Box/Box';

// Screens for media queries
import screens from '../_utils/globals/screens';

const colWidths = {
	1: 8.33333333,
	2: 16.66666667,
	3: 25,
	4: 33.33333333,
	5: 41.66666667,
	6: 50,
	7: 58.33333333,
	8: 66.66666667,
	9: 75,
	10: 83.33333333,
	11: 91.66666667,
	12: 100,
};

const generateMediaQuery = (media: string, colWidth?: ColWidth) => {
	if (colWidth) {
		return `
            @media ${media} {
                flex-grow: ${colWidth};
                flex-shrink: 1;
                width: ${colWidths[colWidth]}%;
                max-width: ${colWidths[colWidth]}%;
            }
        `;
	}
	return `
        @media ${media} {
            flex-grow: 1;
            flex-shrink: 1;
        }
    `;
};

const generateHiddenMediaQuery = (media: string) => {
	return `
        @media ${media} {
            display: none;
        }
    `;
};

const assembleAvengers = (data: IColWidths) => {
	const { lg, md, sm, xs } = data;

	if (lg || md || sm || xs) {
		const large = lg === 'hidden' ? generateHiddenMediaQuery(screens.large) : generateMediaQuery(screens.large, lg);
		const medium = md === 'hidden' ? generateHiddenMediaQuery(screens.medium) : generateMediaQuery(screens.medium, md);
		const small = sm === 'hidden' ? generateHiddenMediaQuery(screens.small) : generateMediaQuery(screens.small, sm);
		const extraSmall = xs === 'hidden' ? generateHiddenMediaQuery(screens.xSmall) : generateMediaQuery(screens.xSmall, xs);
		return `${large} ${medium} ${small} ${extraSmall}`;
	}
	return `
        flex-grow: 1;
        flex-shrink: 1;
    `;
};

export const StyledGrid = styled(Box)<IGridProps>`
	display: flex;
	flex-direction: column;
	flex-grow: 1;
	flex-shrink: 1;
	box-sizing: border-box;
`;

StyledGrid.displayName = 'StyledGrid';

export const StyledRow = styled(Box)<IRowProps>`
	display: flex;
	flex-flow: row wrap;
	box-sizing: border-box;
	min-width: 0px;
`;

StyledRow.displayName = 'StyledRow';

export const StyledCol = styled(Box)<IColProps>`
	box-sizing: border-box;
	min-width: 0px;
	${(props) => assembleAvengers({ lg: props.lg, md: props.md, sm: props.sm, xs: props.xs })}
`;

StyledCol.displayName = 'StyledCol';
