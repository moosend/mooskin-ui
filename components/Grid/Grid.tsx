import * as React from 'react';

// Mooskin Context HoC that passes context to component props
import { withMooskinContext } from '../Styled/MooskinContextProvider';

// Models
import { IColProps, IGridProps, IRowProps } from './model';

// Styled Components
import { StyledCol, StyledGrid, StyledRow } from './styles';

/**
 * Grid
 */
export const Grid: React.FC<IGridProps> = withMooskinContext((props) => {
	return <StyledGrid {...props} />;
});

Grid.defaultProps = {
	className: '',
	style: {},
};

Grid.displayName = 'Grid';

/**
 * Row
 */
export const Row: React.FC<IRowProps> = withMooskinContext((props) => {
	return <StyledRow {...props} />;
});

Row.defaultProps = {
	className: '',
	style: {},
};

Row.displayName = 'Row';

/**
 * Col
 */
export const Col: React.FC<IColProps> = withMooskinContext((props) => {
	return <StyledCol {...props} />;
});

Col.defaultProps = {
	className: '',
	style: {},
};

Col.displayName = 'Col';
