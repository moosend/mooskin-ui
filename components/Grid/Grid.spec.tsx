import * as React from 'react';
import { Grid, Col, Row } from './Grid';

import { mount } from 'enzyme';

describe('Grid', () => {
	test('renders Grid correctly', () => {
		const tree = mount(
			<Grid style={{ textAlign: 'center' }}>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col lg={12} md={8} sm={6} xs={3} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-12 md-8 sm-6 xs-3
					</Col>
				</Row>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col lg={6} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-6
					</Col>
					<Col lg={6} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-6
					</Col>
				</Row>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col md={6} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col md-6
					</Col>
					<Col md={6} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col md-6
					</Col>
				</Row>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col md={4} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 1
					</Col>
					<Col md={4} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 2
					</Col>
					<Col md={4} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 3
					</Col>
					<Col md={4} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 4
					</Col>
					<Col md={4} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 5
					</Col>
					<Col md={4} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 6
					</Col>
					<Col md={6} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 7
					</Col>
					<Col md={6} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Chart 8
					</Col>
				</Row>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col lg={6} md={12} sm={8} xs={4} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-6 md-12 sm-8 xs-4
					</Col>
					<Col lg={6} md={12} sm={4} xs={8} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-6 md-12 sm-4 xs-8
					</Col>
				</Row>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col lg={4} md={10} sm={1} xs={2} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-4 md-10 sm-1 xs-1
					</Col>
					<Col lg={4} md={1} sm={10} xs={2} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-4 md-1 sm-10 xs-1
					</Col>
					<Col lg={4} md={1} sm={1} xs={8} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-4 md-1 sm-1 xs-10
					</Col>
				</Row>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col lg={8} md={4} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-8 md-4 sm-6 xs-12
					</Col>
					<Col lg={4} md={8} sm={6} xs={12} style={{ padding: 10, border: '3px solid #5ccdde' }}>
						Col lg-4 md-8 sm-6 xs-12
					</Col>
				</Row>
				<Row style={{ padding: 10, border: '3px solid #F48770' }}>
					<Col>
						<Row style={{ padding: 10, border: '3px solid #F48770' }}>
							<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
							<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
							<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
						</Row>
						<Row style={{ padding: 10, border: '3px solid #F48770' }}>
							<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
							<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
						</Row>
						<Row style={{ padding: 10, border: '3px solid #F48770' }}>
							<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
						</Row>
					</Col>
					<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
					<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
					<Col style={{ padding: 10, border: '3px solid #5ccdde' }}>Col</Col>
				</Row>
			</Grid>
		);

		expect(tree).toMatchSnapshot();
	});
});
