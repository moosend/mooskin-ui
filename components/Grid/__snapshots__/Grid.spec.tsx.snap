// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders Col correctly 1`] = `
<div
  className="col-component col large-8 medium-6 small-4 hidden-xs myClass"
  style={
    Object {
      "width": "50px",
    }
  }
/>
`;

exports[`Grid renders Grid correctly 1`] = `
<div
  className="grid-component grid myClass"
  style={
    Object {
      "width": "50px",
    }
  }
/>
`;

exports[`Grid renders Grid with Row & Col correctly 1`] = `
<Grid
  className="myClass"
  style={
    Object {
      "width": "50px",
    }
  }
>
  <div
    className="grid-component grid myClass"
    id={undefined}
    style={
      Object {
        "width": "50px",
      }
    }
  >
    <Row
      className="myClass"
      style={
        Object {
          "width": "50px",
        }
      }
    >
      <div
        className="row-component row myClass"
        id={undefined}
        style={
          Object {
            "width": "50px",
          }
        }
      >
        <Col
          className="myClass"
          lg={8}
          md={6}
          sm={4}
          style={
            Object {
              "width": "50px",
            }
          }
          xs="hidden"
        >
          <div
            className="col-component col large-8 medium-6 small-4 hidden-xs myClass"
            id={undefined}
            style={
              Object {
                "width": "50px",
              }
            }
          >
            Col
          </div>
        </Col>
      </div>
    </Row>
  </div>
</Grid>
`;

exports[`Grid renders Row correctly 1`] = `
<div
  className="row-component row myClass"
  style={
    Object {
      "width": "50px",
    }
  }
/>
`;
